SDK: Add import alias for style and javascript

dmsnell opened the pull request:
Up until now when working with the SDK script we have had to provide
import paths directly to the files we are importing, like this...

```js
import Button from 'components/button';
```

This is problematic especially when the project being built with the
SDK doesn't reside inside the Calypso repository because those relative
paths no longer work and we can't enforc any absolute paths on it
either.

In this PR we're adding new alias configurations in webpack to allow for
a path-independent way of importing code from Calypso.

```js
import Button from '@calypso/components/button';
```

**Testing**

Create a test file in some directory outside of the Calypso repository.
Create both a stylesheet and import it in some JavaScript file.
Use the `gutenberg` target to test with.

```js
cd ~/wp-calypso
npm link
cd ~/other-project
echo "import './style.scss'" > index.js
echo "@import '@calypso-style/_colors.scss'; .test { color: $alert-red; }" > style.scss
calypso-sdk gutenberg --editor-script=$(pwd)/index.js
```

In the output you should see the actual color in the produced stylesheet and in the produced RTL stylesheet.

dmsnell committed:
SDK: Add import alias for style and javascript

Up until now when working with the SDK script we have had to provide
import paths directly to the files we are importing, like this...

```js
import '../../assets/stylesheets/shared/_colors.scss';
import Button from 'components/button';
```

This is problematic especially when the project being built with the
SDK doesn't reside inside the Calypso repository because those relative
paths no longer work and we can't enforc any absolute paths on it
either.

In this PR we're adding new alias configurations in webpack to allow for
a path-independent way of importing code from Calypso.

```js
import '@calypstyle/_colors.scss';
import Button from '@calypso/components/button';
```

Copons commented on webpack.config.js:
```diff
@@ -244,6 +244,8 @@ function getWebpackConfig( { cssFilename, externalizeWordPressPackages = false }
 					'react-virtualized': 'react-virtualized/dist/commonjs',
 					'social-logos/example': 'social-logos/build/example',
 					debug: path.resolve( __dirname, 'node_modules/debug' ),
+					'@calypstyle': path.resolve( path.join( __dirname, 'assets', 'stylesheets', 'shared' ) ),
```
Don't forget to pick an appropriate name for this! 😄 
`@calypso-style` sounds better, but we should probably be more explicit in the fact that it targets the `shared` folder, and not any scss file.
What about `@calypso-style-shared` or `@calypso-style-common`? Are they too much of a mouthful?

Copons commented:
Tested and it works as expected! 🎉 

---

I've got a (minor) question though: SASS allows omitting the extension when importing other SASS files.
E.g. `@import '_colors';` is the same as `@import '_colors.scss'`

With this PR though, omitting the extension results in a build error:
E.g. `File to import not found or unreadable: @calypstyle/_colors.`

Do you think there's a way to add support for this?

gwwar commented:
> import '@calypstyle/_colors.scss';

I'm curious if `calypstyle` a typo or intended name here?

dmsnell commented:
> I'm curious if calypstyle a typo or intended name here?

@gwwar I changed it and forgot to update the summary (done now) because @Copons didn't like my clearly-awesome portmanteaux

simison commented:
Noting that shared utils (including colours) are currently globally imported within sass-pipeline in Webpack:
https://github.com/Automattic/wp-calypso/blob/87149d08dd4299add76d5c9860adfad5df3738a4/webpack.config.js#L192-L201

It's another question if we should change that and always require them only when needed. Since they don't output any CSS, are required anyway by over half the sass files and it's kinda regular workflow with sass to have variables and mixins global, it's probably fine to keep them as such.

Rest of the shared CSS (resets, dropdowns, typography) we should probably re-organise in some other way.

So summa summarum, there probably isn't a huge need for `@calypso-style` at the moment. What do you think?

If we decide to include Gutenberg global variables into the mix, that might change a whole lot though!

dmsnell commented:
@simison thanks for the feedback!

> It's another question if we should change that and always require them only when needed.

ooh. a tough one. it doesn't hurt to include them everywhere on build size so that makes sense, but on the other hand I'm leery of two points

 - we see `$wordpress-blue` in a file and have no idea where it came from; I'd rather have an explicit path back to the definitions through an import for helping people understand the code
 - the magic include seems fragile. if we decide to try and compile the CSS differently I guess it will break until someone discovers how it's glued together inside the `webpack.config.js` (yeah I know the alias is magic too but at least it's specifically a string transform and nothing else). I'm not sure if I'm being inconsistent here or not.

> there probably isn't a huge need for @calypso-style at the moment

if we implicitly provide those utils then I agree

> If we decide to include Gutenberg global variables into the mix, that might change a whole lot though!

which leads me back to the value of explicit imports and dependencies. if we leave an audit-trail in the code through imports then stuff like these I don't think will be a problem if and when they come.

dmsnell committed:
Remove shared style alias

not sure what to do about style imports yet and implicit vs. explicit dependencies on the shared global functions

dmsnell commented:
Closing here because we aren't sure where we want to take this and we introduce confusion right now if we start adding the `@` alias without having corresponding `npm` packages

simison commented:
Convo @ p1536340128000100-slack-calypso-gutenberg
